pcapnggo/write.go

```diff
@@ -0,0 +1,392 @@
+		return fmt.Errorf("Can't send statistics for non existent interface %d; have only %d interfaces", intf, w.intf)
+	}
+
+	var scratch [4]option
+	i := 0
+	if !stats.StartTime.IsZero() {
+		scratch[i].code = optionInterfaceStatisticsStartTime
+		scratch[i].raw = stats.StartTime
+		i++
+	}
+	if !stats.EndTime.IsZero() {
+		scratch[i].code = optionInterfaceStatisticsEndTime
+		scratch[i].raw = stats.EndTime
+		i++
+	}
+	if stats.PacketsDropped != NoValue64 {
+		scratch[i].code = optionInterfaceStatisticsInterfaceDropped
+		scratch[i].raw = stats.PacketsDropped
+		i++
+	}
+	if stats.PacketsReceived != NoValue64 {
+		scratch[i].code = optionInterfaceStatisticsInterfaceReceived
+		scratch[i].raw = stats.PacketsReceived
+		i++
+	}
+	options := scratch[:i]
+
+	length := prepareOptions(options) + 24
+
+	ts := stats.LastUpdate.UnixNano()
+	if stats.LastUpdate.IsZero() {
+		ts = 0
+	}
+
+	binary.LittleEndian.PutUint32(w.buf[:4], uint32(blockTypeInterfaceStatistics))
+	binary.LittleEndian.PutUint32(w.buf[4:8], length)
+	binary.LittleEndian.PutUint32(w.buf[8:12], uint32(intf))
+	binary.LittleEndian.PutUint32(w.buf[12:16], uint32(ts>>32))
+	binary.LittleEndian.PutUint32(w.buf[16:20], uint32(ts))
+	if _, err = w.w.Write(w.buf[:20]); err != nil {
+		return
+	}
+
+	if err = w.writeOptions(options); err != nil {
+		return
+	}
+
+	binary.LittleEndian.PutUint32(w.buf[0:4], length)
+	_, err = w.w.Write(w.buf[:4])
+	return
+}
+
+// WritePacket writes out packet with the given data and capture info. The given InterfaceIndex must already be added to the file. InterfaceIndex 0 is automatically added by the NewWriter* methods.
+func (w *Writer) WritePacket(ci gopacket.CaptureInfo, data []byte) (err error) {
+	if ci.InterfaceIndex >= int(w.intf) || ci.InterfaceIndex < 0 {
+		return fmt.Errorf("Can't send statistics for non existent interface %d; have only %d interfaces", ci.InterfaceIndex, w.intf)
+	}
+	if ci.CaptureLength != len(data) {
+		return fmt.Errorf("capture length %d does not match data length %d", ci.CaptureLength, len(data))
+	}
+	if ci.CaptureLength > ci.Length {
+		return fmt.Errorf("invalid capture info %+v:  capture length > length", ci)
+	}
+
+	length := uint32(len(data)) + 32
+	padding := (4 - length&3) & 3
+	length += padding
+
+	ts := ci.Timestamp.UnixNano()
+
+	binary.LittleEndian.PutUint32(w.buf[:4], uint32(blockTypeEnhancedPacket))
+	binary.LittleEndian.PutUint32(w.buf[4:8], length)
+	binary.LittleEndian.PutUint32(w.buf[8:12], uint32(ci.InterfaceIndex))
+	binary.LittleEndian.PutUint32(w.buf[12:16], uint32(ts>>32))
+	binary.LittleEndian.PutUint32(w.buf[16:20], uint32(ts))
+	binary.LittleEndian.PutUint32(w.buf[20:24], uint32(ci.CaptureLength))
+	binary.LittleEndian.PutUint32(w.buf[24:28], uint32(ci.Length))
+
+	if _, err = w.w.Write(w.buf[:28]); err != nil {
+		return
+	}
+
+	if _, err = w.w.Write(data); err != nil {
+		return
+	}
+
+	binary.LittleEndian.PutUint32(w.buf[:4], 0)
+	_, err = w.w.Write(w.buf[4-padding : 8]) // padding + length
```

Comment:
No matter what I did, I couldn't convince go to remove this bounds check - but it's the only one in the hot path, so I think it should be ok.